File: Ch02/ex2_22.erl
-module(ex2_22).
-author("Kevin C. Baird").
-purpose("SICP in Erlang: Exercise 2.22 - make square_list iterative.").
-export([square_list/1]).
-include_lib("eunit/include/eunit.hrl").

square_list(L) -> lists:reverse(square_list(L, [])).

% PRIVATE HELPER FUNCTIONS

square_list([], Answer)    -> Answer;
square_list([H|T], Answer) -> square_list(T, [square(H)|Answer]).

square(X) -> X * X.

%%% TESTS

sq_test() -> ?assert(square_list([0, 1, 2, 3, 4]) =:= [0, 1, 4, 9, 16]).

File: Ch02/ex2_22.exs
# SICP in Elixir: Exercise 2.22 - Make square-list iterative

defmodule MyLists do
  def square(list) do
    iter(list, [])
    |> :lists.reverse()
  end

  # Private functions
  defp iter([], answer), do: answer

  defp iter([hd | tl], answer) do
    iter(tl, [hd * hd] ++ answer)
  end
end

result = MyLists.square([0, 1, 2, 3])
unless result == [0, 1, 4, 9], do: throw(result)
